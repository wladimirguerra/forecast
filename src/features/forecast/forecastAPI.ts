/**
 * All the call to the forecast API is made here in this module.
 *
 * The calls are made to the [Open Weather Map]{@link https://openweathermap.org/forecast5}
 */
import { GeoCoordinates } from "../interfaces";

interface FiveDaysThreeHoursForecastListItem {
  /**
   * Time of the forecasted data. Unix UTC.
   */
  dt: number;
  main: {
    /**
     * Temperature
     * Units – default: kelvin, metric: Celsius, imperial: Fahrenheit.
     * @see {@link ForecastApiRequestQueryParams.units} for options.
     */
    temp: number;
    /**
     * The human perception of weather.
     * Units – default: kelvin, metric: Celsius, imperial: Fahrenheit.
     * @see {@link ForecastApiRequestQueryParams.units} for options.
     */
    feels_like: number;
    /**
     * Minimum daily temperature
     * @see {@link ForecastApiRequestQueryParams.units} about the unit.
     */
    temp_min: number;
    /**
     * Maximum daily temperature
     * @see {@link ForecastApiRequestQueryParams.units} about the unit.
     */
    temp_max: number;
    /**
     * Unit: %
     */
    humidity: number;
  };
  weather: {
    /**
     * Weather condition code
     */
    id: number;
    /**
     * Group of weather parameters
     */
    main: string;
    /**
     * Weather condition within the group
     */
    description: string;
    /**
     * Weather icon id
     * @see [How to get icon URL]{@link https://openweathermap.org/weather-conditions#Weather-Condition-Codes-2}
     */
    icon: string;
  };
}

export interface FiveDaysThreeHoursForecastDataResponse extends GeoCoordinates {
  cod: string;
  message: number;
  cnt: number;
  list: FiveDaysThreeHoursForecastListItem[];
  city: {
    /**
     * Shift in seconds from UTC
     */
    timezone: number;
    name: string;
    country: string;
    coord: GeoCoordinates;
    /**
     * Sunrise time, Unix, UTC
     */
    sunrise: number;
    /**
     * Sunset time, Unix, UTC
     */
    sunset: number;
  };
}

export interface FiveDaysThreeHoursRequestQueryParams extends GeoCoordinates {
  /**
   * Units of measurement.
   *
   * If you do not use the `units` parameter, `standard` units will be applied by default.
   *
   * Temperature is available in Fahrenheit, Celsius and Kelvin units.
   * Wind speed is available in miles/hour and meter/sec.
   *
   * - For temperature in Fahrenheit and wind speed in miles/hour, use units=imperial
   * - For temperature in Celsius and wind speed in meter/sec, use units=metric
   * - Temperature in Kelvin and wind speed in meter/sec is used by default, so there is no need to
   * use the units parameter in the API call if you want this
   */
  units?: "standard" | "metric" | "imperial";
  /**
   * A unique API key.
   *
   * @see ["API key" tab]{@link https://home.openweathermap.org/api_keys} for further details.
   */
  appid: string;
  mode?: string;
  /**
   * A positive integer that limits the number of timestamps in the API response.
   */
  cnt?: number;
  /**
   * Language
   */
  lang?:
    | "af"
    | "al"
    | "ar"
    | "az"
    | "bg"
    | "ca"
    | "cz"
    | "da"
    | "de"
    | "el"
    | "en"
    | "eu"
    | "fa"
    | "fi"
    | "fr"
    | "gl"
    | "he"
    | "hi"
    | "hr"
    | "hu"
    | "id"
    | "it"
    | "ja"
    | "kr"
    | "la"
    | "lt"
    | "mk"
    | "no"
    | "nl"
    | "pl"
    | "pt"
    | "pt_br"
    | "ro"
    | "ru"
    | "sv"
    | "se"
    | "sk"
    | "sl"
    | "sp"
    | "es"
    | "sr"
    | "th"
    | "tr"
    | "ua"
    | "uk"
    | "vi"
    | "zh_cn"
    | "zh_tw"
    | "zu";
}
