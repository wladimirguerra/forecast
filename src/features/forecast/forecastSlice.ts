import { createSlice } from "@reduxjs/toolkit";

export interface DailyForecast {
  day: Date;
  humidity: number;
  dayTemperature: number;
  morningTemperature: number;
  nightTemperature: number;
  meanTemperature: number;
  maximumTemperature: number;
  minimumTemperature: number;
  feelsLikeTemperature: number;
  mode: number;
}

export interface ForecastState {
  daysForecast: DailyForecast[];
}

// Thunks

export const forecastSlice = createSlice({
  name: "forecast",
  initialState: {
    daysForecast: [],
  } as ForecastState,
  reducers: {},
  extraReducers: (/*builder*/) => {},
});

// Selectors

/*export const {
  // TODO fill the actions
} = forecastSlice.actions;*/
export default forecastSlice.reducer;
