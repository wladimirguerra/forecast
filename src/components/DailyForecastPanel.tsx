import * as React from "react";
import { Box, Grid, Paper, Typography } from "@mui/material";
import { Thermostat } from "@mui/icons-material";
import { DailyForecast } from "../features/forecast/forecastSlice";
import moment from "moment";

interface DisplayProps {
  label: string;
  value: number;
  unit: string;
}

const DayIntervalBox: React.FC<DisplayProps> = (props) => {
  const { label, value, unit } = props;

  if (!label) throw new Error("label prop must not be null");

  if (!value) return null;

  return (
    <Grid item xs={3}>
      <Box display={"flex"} flexDirection={"column"} alignItems={"center"}>
        <Typography
          component={"div"}
          variant={"h6"}
          sx={{
            margin: 0,
            lineHeight: 0.9,
          }}
        >
          {value}
          {unit}
        </Typography>
        <Box
          sx={{
            marginTop: 1,
          }}
        >
          <Typography
            variant={"caption"}
            component={"div"}
            color={"text.secondary"}
          >
            {label}
          </Typography>
        </Box>
      </Box>
    </Grid>
  );
};

const DayStatistics: React.FC<DisplayProps> = (props) => {
  const { label, value, unit } = props;
  return (
    <Typography
      component={"span"}
      sx={{
        lineHeight: 1,
        fontSize: "0.8rem",
      }}
      align={"right"}
      color={"secondary"}
      title={label}
    >
      {value}
      {unit}
    </Typography>
  );
};

export interface DailyForecastPanelProps {
  forecast: DailyForecast;
}

const DailyForecastPanel: React.FC<DailyForecastPanelProps> = (props) => {
  const { forecast } = props;

  return (
    <Paper
      sx={{
        borderRadius: 4,
        overflow: "hidden",
        background: "rgba(121,186,255,0.15)",
        display: "flex",
        flexDirection: "column",
        padding: 2,
        maxWidth: {
          sx: "100%",
          sm: 300,
        },
      }}
    >
      <Box display={"flex"} justifyContent={"space-between"}>
        <Box>
          <Typography
            component={"div"}
            variant={"caption"}
            color={"text.secondary"}
          >
            {moment(forecast.day).format("dddd,").toUpperCase()}
          </Typography>
          <Typography variant={"h5"} component={"div"} noWrap>
            {moment(forecast.day).format(" MMMM Do")}
          </Typography>
        </Box>
        <Box
          display={"flex"}
          alignItems={"flex-end"}
          justifyContent={"flex-end"}
        >
          <Thermostat
            color={"secondary"}
            sx={{ marginBottom: 0.4, marginRight: 0.3 }}
          />
          <Typography
            variant={"h3"}
            component={"div"}
            sx={{ lineHeight: 1 }}
            color={"secondary"}
            noWrap
          >
            {forecast.dayTemperature}˚F
          </Typography>
          <Box
            display={"flex"}
            flexDirection={"column"}
            justifyContent={"space-between"}
            alignItems={"center"}
            gap={0.5}
            sx={{
              marginLeft: 1,
            }}
          >
            <DayStatistics
              label={"Max"}
              value={forecast.maximumTemperature}
              unit={"˚"}
            />
            <DayStatistics
              label={"Mean"}
              value={forecast.meanTemperature}
              unit={"˚"}
            />
            <DayStatistics
              label={"Min"}
              value={forecast.minimumTemperature}
              unit={"˚"}
            />
          </Box>
        </Box>
      </Box>
      <Grid container justifyContent={"space-evenly"} sx={{ marginTop: 6 }}>
        <DayIntervalBox
          label={"MORNING"}
          value={forecast.morningTemperature}
          unit={"˚"}
        />
        <DayIntervalBox
          label={"NIGHT"}
          value={forecast.nightTemperature}
          unit={"˚"}
        />
        <DayIntervalBox
          label={"FEELING"}
          value={forecast.feelsLikeTemperature}
          unit={"˚"}
        />
        <DayIntervalBox
          label={"HUMIDITY"}
          value={forecast.humidity}
          unit={"%"}
        />
      </Grid>
    </Paper>
  );
};

export default DailyForecastPanel;
