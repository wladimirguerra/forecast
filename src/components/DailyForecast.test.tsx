import renderer from "react-test-renderer";
import DailyForecastPanel from "./DailyForecastPanel";
import { DailyForecast } from "../features/forecast/forecastSlice";

const FORECAST_MOCK: DailyForecast = {
  day: new Date(2022, 6, 14),
  humidity: 60,
  mode: 10,
  dayTemperature: 21,
  morningTemperature: 7,
  nightTemperature: 10,
  maximumTemperature: 23,
  minimumTemperature: 7,
  feelsLikeTemperature: 19,
  meanTemperature: Math.floor((23 + 7) / 2),
};

describe("DailyForecast Component", () => {
  test("match snapshot", () => {
    const tree = renderer
      .create(<DailyForecastPanel forecast={FORECAST_MOCK} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
