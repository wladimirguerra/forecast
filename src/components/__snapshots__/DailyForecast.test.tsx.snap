// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyForecast Component match snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-kgzhwt-MuiPaper-root"
>
  <div
    className="MuiBox-root css-gg4vpm"
  >
    <div
      className="MuiBox-root css-0"
    >
      <div
        className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
      >
        THURSDAY,
      </div>
      <div
        className="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-1961ak2-MuiTypography-root"
      >
         July 14th
      </div>
    </div>
    <div
      className="MuiBox-root css-1kjvpf0"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeMedium css-1gobhg7-MuiSvgIcon-root"
        data-testid="ThermostatIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15 13V5c0-1.66-1.34-3-3-3S9 3.34 9 5v8c-1.21.91-2 2.37-2 4 0 2.76 2.24 5 5 5s5-2.24 5-5c0-1.63-.79-3.09-2-4zm-4-2V5c0-.55.45-1 1-1s1 .45 1 1v1h-1v1h1v2h-1v1h1v1h-2z"
        />
      </svg>
      <div
        className="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap css-t5epxe-MuiTypography-root"
      >
        21
        ˚F
      </div>
      <div
        className="MuiBox-root css-yys0vv"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiTypography-alignRight css-1lrbjvi-MuiTypography-root"
          title="Max"
        >
          23
          ˚
        </span>
        <span
          className="MuiTypography-root MuiTypography-body1 MuiTypography-alignRight css-1lrbjvi-MuiTypography-root"
          title="Mean"
        >
          15
          ˚
        </span>
        <span
          className="MuiTypography-root MuiTypography-body1 MuiTypography-alignRight css-1lrbjvi-MuiTypography-root"
          title="Min"
        >
          7
          ˚
        </span>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container css-yfgpcs-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-dvxtzn"
      >
        <div
          className="MuiTypography-root MuiTypography-h6 css-tr8st4-MuiTypography-root"
        >
          7
          ˚
        </div>
        <div
          className="MuiBox-root css-164r41r"
        >
          <div
            className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
          >
            MORNING
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-dvxtzn"
      >
        <div
          className="MuiTypography-root MuiTypography-h6 css-tr8st4-MuiTypography-root"
        >
          10
          ˚
        </div>
        <div
          className="MuiBox-root css-164r41r"
        >
          <div
            className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
          >
            NIGHT
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-dvxtzn"
      >
        <div
          className="MuiTypography-root MuiTypography-h6 css-tr8st4-MuiTypography-root"
        >
          19
          ˚
        </div>
        <div
          className="MuiBox-root css-164r41r"
        >
          <div
            className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
          >
            FEELING
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-dvxtzn"
      >
        <div
          className="MuiTypography-root MuiTypography-h6 css-tr8st4-MuiTypography-root"
        >
          60
          %
        </div>
        <div
          className="MuiBox-root css-164r41r"
        >
          <div
            className="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
          >
            HUMIDITY
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
